compiler/rmi/kotlinr/src/KotlinCompilerClient.kt

```diff
@@ -0,0 +1,262 @@
+            val service = connectToService(compilerId, daemonOptions, errStream)
+            if (service != null) {
+                if (!checkId || checkCompilerId(service, compilerId, errStream)) {
+                    errStream.println("[daemon client] found the suitable daemon")
+                    return service
+                }
+                errStream.println("[daemon client] compiler identity don't match: " + compilerId.asParams.joinToString(" "))
+                if (!autostart) return null;
+                errStream.println("[daemon client] shutdown the daemon")
+                service.shutdown()
+                // TODO: find more reliable way
+                Thread.sleep(1000)
+                errStream.println("[daemon client] daemon shut down correctly, restarting")
+            }
+            else {
+                if (!autostart) return null;
+                else errStream.println("[daemon client] cannot connect to Compile Daemon, trying to start")
+            }
+
+            startDaemon(compilerId, daemonLaunchingOptions, daemonOptions, errStream)
+            errStream.println("[daemon client] daemon started, trying to connect")
+            return connectToService(compilerId, daemonOptions, errStream)
+        }
+
+        public fun shutdownCompileService(): Unit {
+            KotlinCompilerClient.connectToCompileService(CompilerId(), DaemonLaunchingOptions(), DaemonOptions(), System.out, autostart = false, checkId = false)
+                    ?.shutdown()
+        }
+
+        public fun incrementalCompile(compiler: CompileService, args: Array<String>, caches: Map<String, IncrementalCache>, out: OutputStream): Int {
+
+            val outStrm = RemoteOutputStreamServer(out)
+            val cacheServers = hashMapOf<String, RemoteIncrementalCacheServer>()
+            try {
+                caches.forEach { cacheServers.put( it.getKey(), RemoteIncrementalCacheServer( it.getValue())) }
+                val res = compiler.remoteIncrementalCompile(args, cacheServers, outStrm, CompileService.OutputFormat.XML)
+                return res
+            }
+            finally {
+                cacheServers.forEach { it.getValue().disconnect() }
+                outStrm.disconnect()
+            }
+        }
+
+        public fun isDaemonEnabled(): Boolean = System.getProperty(COMPILE_DAEMON_ENABLED_PROPERTY) != null
+
+        public fun configureDaemonLaunchingOptions(opts: DaemonLaunchingOptions) {
+            System.getProperty(COMPILE_DAEMON_JVM_OPTIONS_PROPERTY)?.let {
+                // TODO: find better way to pass and parse jvm options for daemon
+                opts.jvmParams = it.split("##")
+            }
+        }
+
+        data class ClientOptions(
+                public var stop: Boolean = false
+        ) :CmdlineParams {
+            override val asParams: Iterable<String>
+                get() =
+                    if (stop) listOf("stop") else listOf()
```

Comment:
Is it now idiomatic? Doesn't look too appealing to me - it is longer and no more expressive for my eyes.